### PR TITLE
Chore: Start Typescript migration

### LINKED ISSUE #786: Typescript
It's a serious amount of work, but we should add Typescript to keep our app maintainable.
Hopefully it won't have any performance impact or new bugs, but we should do it anyway.

React Native supports Typescript without any harm since 0.57.

We need to do it in baby steps to guarantee not to break anything.
- [ ] Setup: it shouldn't be hard, but we need to make sure everything is working afterwards (app, jest, storybook, detox, release mode, ci)
- [ ] Convert OnboardingView: this is one of the smallest views we have and it's the first one in the navigation stack. Easier to migrate and to test
- [ ] Convert other files module by module (Auth, Rooms, Room, etc).

### PR DESCRIPTION
**Pull request by AlexAlexandre**
## Proposed changes
The purpose of this PR is to start the project's migration to typescript.
With that, we need to pay attention to the following points:

- Add a `tsconfig.json` file. In this process, I created this file with the default rules. At the moment, I'm allowing the `type: any` to be used, as this is the beginning of the migration, we still don't know all the types used, in the future it will be necessary to disable this option.
- Update eslint config to support .ts and .tsx files. In this process, was necessary override some js rules, so I created separated rules to .ts. I also changed our extends rules from `airbnb` to `@rocket.chat/eslint-config`, since our company has default lint rules, that's no reason to not use it.
- We have adopted `interfaces` as the default for typing in our repository, following the example below:

```
interface IActionSheetData {
	options: any;
	headerHeight?: number;
	hasCancel?: boolean;
	customHeader: any;
}
```
All interface name needs start with an `I`, as `IActionSheetData `, `IAvatar`, `IButton`... This is a common pattern, and helps all to know when see an interface on the codebase.

- Componentes migrated:
  - `app/constants`;
  - `app/containers`;
  - `app/presentation`;
  - In this process, was necessary migrated some separate files, for example: `app/lib/Navigation.ts` and `app/lib/ShareNavigation.ts`. To use your resources elsewhere.

### Prettier
Taking advantage of all the changes made on `lint`, I also set up `prettier` in the project, to help us with te code formatting.
This tool has very interesting features, and I recommend that all developers involved in this project configure their IDE for automatic code checking and integration.

To more information, please see the link: [Prettier](https://prettier.io/docs/en/index.html).

Thinking about ensuring the quality of the code, I took the liberty of changing our pre-commit script, to not only run lint, but also check the formatting of the code using prettier. The new script now runs this command: 

`"prettier-lint": "yarn lint --fix && prettier --write ."`

## Important
Depends on https://github.com/RocketChat/Rocket.Chat.js.SDK/pull/145
Depends on https://github.com/RocketChat/react-native-ui-lib/pull/3

## Issue(s)	
Closes #786

## How to test or reproduce

We can test the `prettier` following this steps:
To check your code formatting, run this on your terminal:

```sh
yarn prettier-check
```

To fix your code formatting issues, run this on your terminal:

```sh
yarn prettier
```

## Screenshots
N/A

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] Improvement (non-breaking change which improves a current function)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Documentation update (if none of the other choices apply)

## Checklist

- [x] I have read the [CONTRIBUTING](https://github.com/RocketChat/Rocket.Chat/blob/develop/.github/CONTRIBUTING.md#contributing-to-rocketchat) doc
- [x] I have signed the [CLA](https://cla-assistant.io/RocketChat/Rocket.Chat.ReactNative)
- [x] Lint and unit tests pass locally with my changes
- [ ] I have added tests that prove my fix is effective or that my feature works (if applicable)
- [ ] I have added necessary documentation (if applicable)
- [ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: app/containers/ThreadDetails.tsx [hunk 1]
```diff
@@ -40,14 +39,29 @@ const styles = StyleSheet.create({
...
+	item: {
+		tcount: number | string;
+		replies: any;
+		id: string;
+	};
+	user: {
+		id: string
+	};
+	badgeColor: string;
+	toggleFollowThread: Function;
```
Comment (diegolmello) [inline app/containers/ThreadDetails.tsx:52]: What's the difference between `Function` and `(): void`?
You used the void option on `TextInput.tsx`, for example.
We should establish a pattern there.

File: app/containers/markdown/Emoji.tsx [hunk 2]
```diff
@@ -8,11 +7,24 @@ import { themes } from '../../constants/colors';
 
 import styles from './styles';
 
+// TODO this interface it's duplicated with 'app/containers/EmojiPicker/index.tsx'
```
Comment (diegolmello) [inline app/containers/markdown/Emoji.tsx:10]: Kind of duplicated, yes, but it contains a few more props than EmojiPicker.

File: __tests__/__snapshots__/Storyshots.test.js.snap [hunk 3]
```diff
@@ -47791,6 +47791,7 @@ exports[`Storyshots Message With image 1`] = `
                       />
                     </View>
                   </View>
+                  //TODO - fix the required fields for the Markdown
```
[the remark this answers is not in the document]
Comment (AlexAlexandre) [reply __tests__/__snapshots__/Storyshots.test.js.snap:47794]: Nops, 
I add this TODO on `app/containers/message/Image.tsx` line 72. 
This is that case that I already told you, we'll need to analyze calmly the Markdown component, to see all necessary/unnecessary props.

File: __tests__/__snapshots__/Storyshots.test.js.snap [hunk 4]
```diff
@@ -4693,7 +4693,7 @@ exports[`Storyshots List pressable 1`] = `
             "justifyContent": "center",
             "paddingHorizontal": 12,
           },
-          false,
+          undefined,
```
[the remark this answers is not in the document]
Comment (AlexAlexandre) [reply __tests__/__snapshots__/Storyshots.test.js.snap:4696]: I searched for hours where this change came from, I ran several tests, but I couldn't find it.

File: app/containers/Avatar/Avatar.tsx [hunk 5]
```diff
@@ -1,21 +1,17 @@
 import React from 'react';
-import PropTypes from 'prop-types';
 import { View } from 'react-native';
 import FastImage from '@rocket.chat/react-native-fast-image';
 import Touchable from 'react-native-platform-touchable';
 import { settings as RocketChatSettings } from '@rocket.chat/sdk';
 
 import { avatarURL } from '../../utils/avatar';
 import Emoji from '../markdown/Emoji';
+import {TAvatar} from "./types";
```
[the remark this answers is not in the document]
Comment (AlexAlexandre) [reply app/containers/Avatar/Avatar.tsx]: let's use the single quotes, the lint it's already set up to use them

File: app/containers/EmojiPicker/index.tsx [hunk 6]
```diff
@@ -19,21 +18,35 @@ import log from '../../utils/log';
...
 
-class EmojiPicker extends Component {
-	static propTypes = {
-		baseUrl: PropTypes.string.isRequired,
-		customEmojis: PropTypes.object,
-		onEmojiSelected: PropTypes.func,
-		tabEmojiStyle: PropTypes.object,
-		theme: PropTypes.string
-	};
+interface IEmojiPickerProps {
```
[the remark this answers is not in the document]
Comment (AlexAlexandre) [reply app/containers/EmojiPicker/index.tsx:27]: you're right, I forgot to remove this `literal`
already fixed

File: app/containers/RoomHeader/RoomHeader.tsx [hunk 7]
```diff
@@ -41,14 +38,47 @@ const styles = StyleSheet.create({
...
+}
+
+const SubTitle = React.memo(({ usersTyping, subtitle, renderFunc, theme, scale }: TRoomHeaderSubTitle) => {
 	const fontSize = getSubTitleSize(scale);
 	// typing
-	if (usersTyping.length) {
+	if (usersTyping?.length) {
 		let usersText;
-		if (usersTyping.length === 2) {
+		if (usersTyping?.length === 2) {
```
[the remark this answers is not in the document]
Comment (AlexAlexandre) [reply app/containers/RoomHeader/RoomHeader.tsx:81]: Yes! this one I forgot to remove from some tests 
already fixed.

File: app/containers/ThreadDetails.tsx (hunk 1 above)
Comment (AlexAlexandre) [reply app/containers/ThreadDetails.tsx:52]: I don't know the best way to explain that for you, but basically: 
`(): void` : this is a function with more detail about the return (void in this case), and we can pass the number of the parameters that the function will receive. 
`Function`: this one it's more generic, I use that when we receive a function as a parameter, and I did't find more details.

File: app/containers/markdown/Emoji.tsx (hunk 2 above)
Comment (AlexAlexandre) [reply app/containers/markdown/Emoji.tsx:10]: yeah, I'll remove this `TODO` because we already discussed about it.

**Review by diegolmello — APPROVED** [review]
LGTM :)

**Pull request merged**